### PR TITLE
Monitor SageMaker updates

### PR DESCRIPTION
**Pull request by ksaaskil**
Sorry for the huge PR, it took quite some effort to get stuff working.
- Fetch job metrics using [sagemaker.analytics](https://sagemaker.readthedocs.io/en/latest/analytics.html)
- Compute the difference of records in previous and new metrics, assuming that `sagemaker.analytics` appends new rows to the end of DataFrame once new values emerge. **This should be checked**
- Add new scalars to the job via `job.add_scalar`
- Notify updates via the same notify method as for scheduler jobs
- Update example notebook to use a cheaper instance (takes three hours to train now), add some new metrics to be outputted by the script (learning rate and training loss)
- Copy-paste code from `Scheduler` for querying and posting job metrics, these should probably go to `TrackerBase` or `JobMonitorBase` to avoid duplicating code
- Use `threading.Lock` in initializing `SageMakerHelper` connections to avoid race conditions
- Minimum polling interval of 60 secs in `SageMakerJobMonitor`. It only notifies when new values are available, though.
- Change `meeshkan stop` to not raise an exception when used before starting the agent
- Create `terminate_daemon` event in the daemon process, this should be more robust than setting it before spawning and forking (?)
- [x]  Something's broken in Azure tests, I could also reproduce this locally at some point... EDIT: Added longer waiting times and changed `maxParallel` to one. Also changed `terminate_daemon` to be created in the API process. The process can only be stopped via the Pyro proxy anyway in real usage, as only tests re-use the `Service` instance for `start` -> `stop` etc.

### CODE REVIEW

File: meeshkan/core/api.py [hunk 1]
```diff
@@ -197,6 +197,7 @@ def get_updates(self, job_id: uuid.UUID) -> HistoryByScalar:
     @Pyro4.expose
     def stop(self):
         if self.__was_shutdown:
+            LOGGER.warning("Agent API was shutdown twice.")
```
Comment (idantene) [inline meeshkan/core/api.py]: When does this happen?

File: meeshkan/core/sagemaker_monitor.py [hunk 2]
```diff
@@ -91,8 +106,8 @@ def get_job_status(self, job_name) -> JobStatus:
         :raises JobNotFoundException: If job was not found.
         :return: Job status
         """
-
-        self.check_or_build_connection()
+        with self.lock:
```
Comment (idantene) [inline meeshkan/core/sagemaker_monitor.py]: Why a thread lock? Are there any race conditions?

File: meeshkan/core/sagemaker_monitor.py [hunk 3]
```diff
@@ -195,6 +280,24 @@ def start(self, job: SageMakerJob) -> asyncio.Task:
             LOGGER.info("Stopped monitoring SageMakerJob %s, got status %s", job.name, job.status)
         except asyncio.CancelledError:
             LOGGER.debug("SageMakerJob tracking cancelled for job %s", job.name)
+        except Exception:  # pylint:disable=broad-except
+            LOGGER.exception("Polling for updates failed")
+            # Ignore
+
+    # TODO Remove duplicate code with Scheduler
```
Comment (idantene) [inline meeshkan/core/sagemaker_monitor.py]: Goes back to perhaps having a common base class with `Scheduler` :thinking:

File: meeshkan/core/service.py [hunk 4]
```diff
@@ -66,6 +67,8 @@ def daemonize(self, serialized):
         pid = os.fork()
         if pid > 0:  # Close parent process
             return
+        if not self.terminate_daemon:
+            self.terminate_daemon = multiprocessing.get_context("spawn").Event()
```
Comment (idantene) [inline meeshkan/core/service.py]: `multiprocessing.get_context("spawn")` should also be abstracted away somehow so we don't duplicate it everywhere
